apps/prairielearn/src/ee/lib/lti13.ts

```diff
@@ -499,3 +541,186 @@ export async function fetchRetry(
     return await fetchRetry(input, opts, fetchRetryOpts);
   }
 }
+
+// https://www.imsglobal.org/spec/lti-ags/v2p0#score-publish-service
+export const Lti13ScoreSchema = z.object({
+  scoreGiven: z.number(),
+  scoreMaximum: z.number(),
+  userId: z.string(),
+  scoringUserId: z.string().optional(),
+  activityProgress: z.enum(['Initialized', 'Started', 'InProgress', 'Submitted', 'Completed']),
+  gradingProgress: z.enum(['FullyGraded', 'Pending', 'PendingManual', 'Failed', 'NotReady']),
+  timestamp: DateFromISOString,
+  submission: z.any().optional(),
+  startedAt: DateFromISOString.optional(),
+  submittedAt: DateFromISOString.optional(),
+  comment: z.string().optional(),
+});
+export type Lti13Score = z.infer<typeof Lti13ScoreSchema>;
+
+const UsersWithLti13SubSchema = UserSchema.extend({
+  lti13_sub: z.string().nullable(),
+});
+type UsersWithLti13Sub = z.infer<typeof UsersWithLti13SubSchema>;
+
+// https://www.imsglobal.org/spec/lti-nrps/v2p0/#sharing-of-personal-data
+const ContextMembershipSchema = z.object({
+  user_id: z.string(),
+  roles: z.string().array(), // https://www.imsglobal.org/spec/lti/v1p3#role-vocabularies
+  status: z.enum(['Active', 'Inactive']).optional(),
+  email: z.string(), // optional per spec but required for us
+});
+type ContextMembership = z.infer<typeof ContextMembershipSchema>;
+
+class Lti13ContextMembership {
```

Comment:
I tweaked the API + implementation to be a little safer and more efficient in ab9532e.

Reply:
I'm OK with most of the changes. What sticks out as kind of odd to me is a class with a static function that returns a new class instance, that also has a constructor. It seems more straightforward to me to either have the class without a constructor and a public function `loadForInstance` that initializes. Or put all of the `loadForInstance` logic into a constructor?

It's not a big deal, since we're only calling this in one context and can refactor it later. I think to me the hiding the `new` in a static function makes it feel less class-like?

Reply:
That's effectively a workaround for the fact that a class can't have an async constructor. An API where you _must_ call some kind of function to initialize internal state before the class instance is usable feels weirder to me than this.

This of course doesn't at all have to be a class, it could just as easily be functions:

```ts
async function loadMemberships() {
  const memberships = await doNetworkStuff();

  return {
    lookup(user) {
      // look up in `memberships`
    }
  }
}
```

I don't have a strong preference either way.